compiler/rustc_const_eval/src/interpret/intrinsics.rs

```diff
@@ -413,35 +413,41 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                         ),
                     )?;
                 }
-                if intrinsic_name == sym::assert_zero_valid
-                    && !layout.might_permit_raw_init(
-                        self,
-                        InitKind::Zero,
-                        self.tcx.sess.opts.debugging_opts.strict_init_checks,
-                    )
-                {
-                    M::abort(
-                        self,
-                        format!(
-                            "aborted execution: attempted to zero-initialize type `{}`, which is invalid",
-                            ty
-                        ),
-                    )?;
+
+                if intrinsic_name == sym::assert_zero_valid {
+                    let should_panic = if self.tcx.sess.opts.debugging_opts.strict_init_checks {
```

Comment:
So each backend needs to implement that `if self.tcx.sess.opts.debugging_opts.strict_init_checks` separately? That doesn't seem great, can it be avoided? (Also cranelift doesn't seem to implement it at all currently?)

Reply:
cg_clif does: https://github.com/bjorn3/rustc_codegen_cranelift/blob/a298c6908e443fc689c474750df26d82301fd74f/src/intrinsics/mod.rs#L679, https://github.com/bjorn3/rustc_codegen_cranelift/blob/a298c6908e443fc689c474750df26d82301fd74f/src/intrinsics/mod.rs#L695

Reply:
I added a dependency on `rustc_const_eval` to cranelift (so it can use the const eval checks), and then now it's `rustc_const_eval::might_permit_raw_init`'s job to defer to the existing checks if strict checks weren't requested.